pkg/volume/aws_ebs/attacher.go

```diff
@@ -0,0 +1,258 @@
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type awsElasticBlockStoreAttacher struct {
+	host volume.VolumeHost
+}
+
+var _ volume.Attacher = &awsElasticBlockStoreAttacher{}
+
+var _ volume.AttachableVolumePlugin = &awsElasticBlockStorePlugin{}
+
+// Singleton key mutex for keeping attach/detach operations for the same PD atomic
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *awsElasticBlockStorePlugin) NewAttacher() (volume.Attacher, error) {
+	return &awsElasticBlockStoreAttacher{host: plugin.host}, nil
+}
+
+func (plugin *awsElasticBlockStorePlugin) GetDeviceName(spec *volume.Spec) (string, error) {
+	volumeSource, _ := getVolumeSource(spec)
+	if volumeSource == nil {
+		return "", fmt.Errorf("Spec does not reference an EBS volume type")
+	}
+
+	return volumeSource.VolumeID, nil
+}
+
+func (plugin *awsElasticBlockStorePlugin) GetUniqueVolumeName(spec *volume.Spec) (string, error) {
+	volumeSource, _ := getVolumeSource(spec)
+	if volumeSource == nil {
+		return "", fmt.Errorf("Spec does not reference an EBS volume type")
+	}
+	partition := ""
+	if volumeSource.Partition != 0 {
+		partition = strconv.Itoa(int(volumeSource.Partition))
+	}
+	return fmt.Sprintf("%s/%s-%s:%v", awsElasticBlockStorePluginName, volumeSource.VolumeID, partition, volumeSource.ReadOnly), nil
+}
+
+func (attacher *awsElasticBlockStoreAttacher) Attach(spec *volume.Spec, hostName string) error {
+	volumeSource, readOnly := getVolumeSource(spec)
+	VolumeID := volumeSource.VolumeID
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(VolumeID)
+	defer attachDetachMutex.UnlockKey(VolumeID)
+
+	awsCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return err
+	}
+
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		if numRetries > 0 {
+			glog.Warningf("Retrying attach for AWS PD %q (retry count=%v).", VolumeID, numRetries)
+		}
+
+		if _, err = awsCloud.AttachDisk(VolumeID, hostName, readOnly); err != nil {
+			glog.Errorf("Error attaching PD %q: %+v", VolumeID, err)
+			time.Sleep(errorSleepDuration)
+			continue
+		}
+		return nil
+	}
+
+	return err
+}
+
+// TODO: use metadata service instead of querying cloud provider to retrieve devicePath
+func (attacher *awsElasticBlockStoreAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	awsCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return "", err
+	}
+	volumeSource, _ := getVolumeSource(spec)
+	VolumeID := volumeSource.VolumeID
+	partition := ""
+	if volumeSource.Partition != 0 {
+		partition = strconv.Itoa(int(volumeSource.Partition))
+	}
+	devicePath, err := awsCloud.GetDiskPath(VolumeID)
+	if err != nil {
+		return "", err
+	}
```

Comment:
@saad-ali if disk is not attached yet, it should return an error. Do you want kubelet or plugin to retry in this case?